hedera-node/src/main/java/com/hedera/services/txns/schedule/ScheduleProcessing.java

```diff
@@ -0,0 +1,267 @@
+import com.hedera.services.store.schedule.ScheduleStore;
+import com.hedera.services.throttling.TimedFunctionalityThrottling;
+import com.hedera.services.throttling.annotations.ScheduleThrottle;
+import com.hedera.services.utils.EntityNum;
+import com.hedera.services.keys.InHandleActivationHelper;
+import com.hedera.services.utils.accessors.PlatformTxnAccessor;
+import com.hedera.services.utils.accessors.SignedTxnAccessor;
+import com.hedera.services.utils.accessors.TxnAccessor;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.ScheduleID;
+import com.swirlds.common.system.transaction.SwirldTransaction;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import static com.hedera.services.utils.EntityNum.fromScheduleId;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.SCHEDULE_FUTURE_GAS_LIMIT_EXCEEDED;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.SCHEDULE_FUTURE_THROTTLE_EXCEEDED;
+
+/**
+ * Class that encapsulates some of the more complex processing of scheduled transactions.
+ */
+@Singleton
+public class ScheduleProcessing {
+	private static final Logger log = LogManager.getLogger(ScheduleProcessing.class);
+
+	private final SigImpactHistorian sigImpactHistorian;
+	private final ScheduleStore store;
+	private final ScheduleExecutor scheduleExecutor;
+	private final GlobalDynamicProperties dynamicProperties;
+	private final SigsAndPayerKeyScreen sigsAndPayerKeyScreen;
+	private final CharacteristicsFactory characteristics;
+	private final TimedFunctionalityThrottling scheduleThrottling;
+
+	SigMapScheduleClassifier classifier = new SigMapScheduleClassifier();
+	SignatoryUtils.ScheduledSigningsWitness signingsWitness = SignatoryUtils::witnessScoped;
+	BiPredicate<ScheduleVirtualValue, InHandleActivationHelper> isReady = SignatoryUtils::isReady;
+
+	@Inject
+	public ScheduleProcessing(final SigImpactHistorian sigImpactHistorian, final ScheduleStore store,
+			final ScheduleExecutor scheduleExecutor, final GlobalDynamicProperties dynamicProperties,
+			final SigsAndPayerKeyScreen sigsAndPayerKeyScreen, final CharacteristicsFactory characteristics,
+			@ScheduleThrottle final TimedFunctionalityThrottling scheduleThrottling) {
+		this.sigImpactHistorian = sigImpactHistorian;
+		this.store = store;
+		this.scheduleExecutor = scheduleExecutor;
+		this.dynamicProperties = dynamicProperties;
+		this.sigsAndPayerKeyScreen = sigsAndPayerKeyScreen;
+		this.characteristics = characteristics;
+		this.scheduleThrottling = scheduleThrottling;
+	}
+
+	/**
+	 * Expires all scheduled transactions, that are in a final state, having an expiry before consensusTime,
+	 * and having an expiry before any transaction that is ready to execute.
+	 */
+	public void expire(Instant consensusTime) {
+
+		while (true) {
```

Comment:
I don't understand why this particular `while (true) {...` is necessary. 🤔  But even if it is, for our collective sanity, I would **strongly** prefer we don't have any potentially unbounded iteration of any kind in `handleTransaction`. 😁

Reply:
That's why I added the `(next.equals(previousId))` check :-). I will update with a little more checks. As long as the db works it will not loop forever. 

We need the loop so we expire all non-executable txns until we get to one that is executable.

Reply:
fixed